File: implementations/java/src/main/java/com/axisrobo/aep/McpBridge.java
package com.axisrobo.aep;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.BiFunction;

public class McpBridge {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    @FunctionalInterface
    public interface EventSink {
        void send(Map<String, Object> event);
    }

    public interface ToolHandler {
        Map<String, Object> handle(Map<String, Object> args, Map<String, Object> ctx);
    }

    public record ToolDefinition(String name, Map<String, Object> schema, ToolHandler handler) {}

    private record Tool(Map<String, Object> schema, ToolHandler handler) {}

    private final EventSink sink;
    private final Harness harness;
    private final Map<String, Tool> tools = new LinkedHashMap<>();
    private final Map<String, Object> serverInfo = Map.of("name", "aep-mcp-bridge", "version", "0.1.0");

    public McpBridge(EventSink sink) {
        this.sink = sink;
        this.harness = new Harness();
    }

    public McpBridge registerTool(String name, Map<String, Object> schema, ToolHandler handler) {
        tools.put(name, new Tool(schema, handler));
        return this;
    }

    public McpBridge registerTool(ToolDefinition def) {
        tools.put(def.name(), new Tool(def.schema(), def.handler()));
        return this;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> handleRequest(Map<String, Object> request) {
        var method = request.get("method") instanceof String s ? s : null;
        if (method == null || method.isEmpty()) {
            return errorResponse(null, -32600, "Invalid Request");
        }
        switch (method) {
            case "initialize": return handleInitialize(request);
            case "notifications/initialized": return null;
            case "tools/list": return handleToolsList(request);
            case "tools/call": return handleToolsCall(request);
            default: return errorResponse(request.get("id"), -32601, "Method not found: " + method);
        }
    }

    private Map<String, Object> handleInitialize(Map<String, Object> request) {
        return Map.of(
            "jsonrpc", "2.0",
            "id", request.getOrDefault("id", null),
            "result", Map.of(
                "protocolVersion", "0.1.0",
                "capabilities", Map.of("tools", Map.of()),
                "serverInfo", serverInfo
            )
        );
    }

    private Map<String, Object> handleToolsList(Map<String, Object> request) {
        var toolList = new ArrayList<Map<String, Object>>();
        for (var entry : tools.entrySet()) {
            var schema = entry.getValue().schema();
            var description = schema.get("description") instanceof String s ? s : "AEP-backed tool: " + entry.getKey();
            var properties = schema.getOrDefault("properties", Map.of());
            var required = schema.getOrDefault("required", List.of());
            toolList.add(Map.of(
                "name", entry.getKey(),
                "description", description,
                "inputSchema", Map.of("type", "object", "properties", properties, "required", required)
            ));
        }
        return Map.of("jsonrpc", "2.0", "id", request.getOrDefault("id", null), "result", Map.of("tools", toolList));
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> handleToolsCall(Map<String, Object> request) {
        var params = request.get("params") instanceof Map<?, ?> p ? (Map<String, Object>) p : Map.<String, Object>of();
        var name = params.get("name") instanceof String s ? s : null;
        var tool = tools.get(name);
        if (tool == null) {
            return errorResponse(request.get("id"), -32602, "Unknown tool: " + name);
        }
        var args = params.get("arguments") instanceof Map<?, ?> a ? (Map<String, Object>) a : Map.<String, Object>of();
        var ctx = new LinkedHashMap<String, Object>();
        ctx.put("harness", harness);
        ctx.put("sink", sink);
        ctx.put("task_id", args.get("_task_id"));
        ctx.put("session_id", args.get("_session_id"));
        try {
            var result = tool.handler().handle(args, ctx);
            return Map.of("jsonrpc", "2.0", "id", request.getOrDefault("id", null), "result", result);
        } catch (Exception err) {
            return Map.of("jsonrpc", "2.0", "id", request.getOrDefault("id", null), "result", Map.of(
                "isError", true,
                "content", List.of(Map.of("type", "text", "text", String.valueOf(err.getMessage())))
            ));
        }
    }

    private Map<String, Object> errorResponse(Object id, int code, String message) {
        var resp = new LinkedHashMap<String, Object>();
        resp.put("jsonrpc", "2.0");
        resp.put("id", id);
        resp.put("error", Map.of("code", code, "message", message));
        return resp;
    }

    public static ToolDefinition asyncToolHandler(String name, String description,
                                                  BiFunction<Map<String, Object>, TaskTracker, Map<String, Object>> work) {
        ToolHandler handler = (args, ctx) -> {
            var taskId = args.get("_task_id") instanceof String s && !s.isEmpty()
                ? s : "task_" + UUID.randomUUID().toString().replace("-", "");
            var tracker = new TaskTracker(taskId, "tool:" + name);
            var sink = ctx.get("sink") instanceof EventSink es ? es : null;

            send(sink, tracker.accepted());

            var thread = new Thread(() -> {
                try {
                    send(sink, tracker.started());
                    send(sink, tracker.progress(Map.of("progress", 0.5, "message", name + " in progress")));
                    var result = work.apply(args, tracker);
                    send(sink, tracker.completed(result));
                } catch (Exception err) {
                    send(sink, tracker.failed(Errors.TOOL_ERROR, String.valueOf(err.getMessage())));
                }
            });
            thread.setDaemon(true);
            thread.start();

            String acceptedText;
            try {
                acceptedText = MAPPER.writeValueAsString(Map.of("task_id", tracker.id(), "status", "accepted"));
            } catch (Exception e) {
                acceptedText = "{\"task_id\":\"" + tracker.id() + "\",\"status\":\"accepted\"}";
            }
            return Map.of("content", List.of(Map.of("type", "text", "text", acceptedText)));
        };
        return new ToolDefinition(name, Map.of("description", description), handler);
    }

    private static void send(EventSink sink, Map<String, Object> event) {
        if (sink != null) sink.send(event);
    }
}

File: implementations/go/mcp-bridge/mcp_bridge.go
package mcpbridge

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/axisrobo/harmovela/event"
	"github.com/axisrobo/harmovela/harness"
	"github.com/axisrobo/harmovela/task"
)

type Sender interface {
	Send(evt map[string]any) error
}

type McpContext struct {
	Harness   *harness.Harness
	Sender    Sender
	TaskID    string
	SessionID string
}

type McpTool struct {
	Name    string
	Schema  map[string]any
	Handler func(args map[string]any, ctx McpContext) map[string]any
}

type McpBridge struct {
	harness    *harness.Harness
	sender     Sender
	tools      map[string]McpTool
	serverInfo map[string]any
}

func NewMcpBridge(sender Sender) *McpBridge {
	return &McpBridge{
		harness:    harness.NewHarness(),
		sender:     sender,
		tools:      make(map[string]McpTool),
		serverInfo: map[string]any{"name": "aep-mcp-bridge", "version": "0.1.0"},
	}
}

func (b *McpBridge) RegisterTool(tool McpTool) *McpBridge {
	b.tools[tool.Name] = tool
	return b
}

func (b *McpBridge) HandleRequest(request map[string]any) map[string]any {
	method, _ := request["method"].(string)
	if method == "" {
		return b.errorResponse(nil, -32600, "Invalid Request")
	}
	switch method {
	case "initialize":
		return b.handleInitialize(request)
	case "notifications/initialized":
		return nil
	case "tools/list":
		return b.handleToolsList(request)
	case "tools/call":
		return b.handleToolsCall(request)
	default:
		return b.errorResponse(request["id"], -32601, "Method not found: "+method)
	}
}

func (b *McpBridge) handleInitialize(request map[string]any) map[string]any {
	return map[string]any{
		"jsonrpc": "2.0",
		"id":      request["id"],
		"result": map[string]any{
			"protocolVersion": "0.1.0",
			"capabilities":    map[string]any{"tools": map[string]any{}},
			"serverInfo":      b.serverInfo,
		},
	}
}

func (b *McpBridge) handleToolsList(request map[string]any) map[string]any {
	tools := make([]map[string]any, 0, len(b.tools))
	for name, tool := range b.tools {
		description, _ := tool.Schema["description"].(string)
		if description == "" {
			description = "AEP-backed tool: " + name
		}
		properties, ok := tool.Schema["properties"]
		if !ok {
			properties = map[string]any{}
		}
		required, ok := tool.Schema["required"]
		if !ok {
			required = []any{}
		}
		tools = append(tools, map[string]any{
			"name":        name,
			"description": description,
			"inputSchema": map[string]any{
				"type":       "object",
				"properties": properties,
				"required":   required,
			},
		})
	}
	return map[string]any{"jsonrpc": "2.0", "id": request["id"], "result": map[string]any{"tools": tools}}
}

func (b *McpBridge) handleToolsCall(request map[string]any) map[string]any {
	params, _ := request["params"].(map[string]any)
	if params == nil {
		params = map[string]any{}
	}
	name, _ := params["name"].(string)
	tool, ok := b.tools[name]
	if !ok {
		return b.errorResponse(request["id"], -32602, "Unknown tool: "+name)
	}
	args, _ := params["arguments"].(map[string]any)
	if args == nil {
		args = map[string]any{}
	}
	sessionID, _ := args["_session_id"].(string)
	taskID, _ := args["_task_id"].(string)
	result := invokeTool(tool, args, McpContext{
		Harness:   b.harness,
		Sender:    b.sender,
		TaskID:    taskID,
		SessionID: sessionID,
	})
	return map[string]any{"jsonrpc": "2.0", "id": request["id"], "result": result}
}

func invokeTool(tool McpTool, args map[string]any, ctx McpContext) (result map[string]any) {
	defer func() {
		if r := recover(); r != nil {
			result = map[string]any{
				"isError": true,
				"content": []any{map[string]any{"type": "text", "text": fmt.Sprintf("%v", r)}},
			}
		}
	}()
	return tool.Handler(args, ctx)
}

func (b *McpBridge) errorResponse(id any, code int, message string) map[string]any {
	return map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"error":   map[string]any{"code": code, "message": message},
	}
}

func AsyncToolHandler(name, description string, work func(args map[string]any, tracker *task.Tracker) map[string]any) McpTool {
	return McpTool{
		Name:   name,
		Schema: map[string]any{"description": description},
		Handler: func(args map[string]any, ctx McpContext) map[string]any {
			taskID, _ := args["_task_id"].(string)
			if taskID == "" {
				taskID = fmt.Sprintf("task_%d", time.Now().UnixMilli())
			}
			tracker := task.NewTracker(taskID, "tool:"+name, jsonString(args))
			send := func(evt map[string]any) {
				if ctx.Sender != nil {
					ctx.Sender.Send(evt)
				}
			}
			send(tracker.Accepted())

			go func() {
				defer func() {
					if r := recover(); r != nil {
						send(tracker.Failed(event.ErrorCodeToolError, fmt.Sprintf("%v", r)))
					}
				}()
				send(tracker.Started())
				send(tracker.Progress(map[string]any{"progress": 0.5, "message": name + " in progress"}))
				result := work(args, tracker)
				send(tracker.Completed(result))
			}()

			acceptedText := jsonString(map[string]any{"task_id": tracker.ID, "status": "accepted"})
			return map[string]any{"content": []any{map[string]any{"type": "text", "text": acceptedText}}}
		},
	}
}

func jsonString(v any) string {
	data, err := json.Marshal(v)
	if err != nil {
		return "{}"
	}
	return string(data)
}
